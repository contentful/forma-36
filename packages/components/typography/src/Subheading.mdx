---
title: 'Subheading'
type: 'component'
status: 'stable'
section: 'typographyComponents'
slug: /components/subheading/
github: 'https://github.com/contentful/forma-36/tree/forma-v4/packages/components/typography/src'
typescript: ./Subheading.tsx
---

`Subheading` is intended to be used as a heading for copy nested within a section. For example: a subheading under Contact may be Address or Phone number.

### Import

```js static=true
import { Subheading } from '@contentful/f36-components';
```

## Examples

### Basic

```jsx file=../examples/SubheadingExample.tsx

```

### Margins

By default `Subheading` component has a margin bottom of `spacingM` (`1rem`). You can override the default margin bottom by specifying it explicitly on the component level by `marginBottom="none"` or any other value from [our spacing options](/tokens/spacing)

```jsx file=../examples/SubheadingSpacingExample.tsx

```

### `as` property

By default `Subheading` has an `h3` tag. But you can change it be providing `as` prop with values from `h1` to `h6`.

```jsx file=../examples/SubheadingTagExample.tsx

```

### Truncated text

Sometimes you might have to truncate the text in the `Subheading` component, so we provide the prop especially for that. Under the hood, what it does is add `overflow: hidden` and set `text-overflow` to `ellipsis`. Have a look at the example below:

```jsx file=../examples/SubheadingTruncatedExample.tsx

```

### Composition

```jsx file=../examples/SubheadingCompositionExample.tsx

```

## Props (API reference)

<Props of="Subheading" storybookPath="/story/typography-subheading--basic" />
