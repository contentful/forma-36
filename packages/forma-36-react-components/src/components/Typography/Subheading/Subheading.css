@import 'settings/colors';
@import 'settings/typography';
@import 'settings/transitions';

.Subheading {
  display: block;
  font-family: var(--font-stack-primary);
  font-weight: var(--font-weight-demi-bold);
  color: var(--color-text-dark);
  font-size: var(--font-size-l);
  line-height: var(--line-height-l);
  margin: 0;
}
