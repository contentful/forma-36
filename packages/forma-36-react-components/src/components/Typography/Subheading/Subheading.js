import React from 'react';
import PropTypes from 'prop-types';
import cn from 'classnames';
import styles from './Subheading.css';

class Subheading extends React.Component {
  static propTypes = {
    extraClassNames: PropTypes.string,
    children: PropTypes.node.isRequired,
    testId: PropTypes.string,
    element: PropTypes.oneOf(['h1', 'h2', 'h3', 'h4', 'h5', 'h6', 'p']),
  };

  static defaultProps = {
    element: 'h2',
    extraClassNames: undefined,
    testId: 'cf-ui-subheading',
  };

  render() {
    const {
      extraClassNames,
      children,
      testId,
      element,
      ...otherProps
    } = this.props;

    const classNames = cn(styles.Subheading, extraClassNames);

    const Element = element;

    return (
      <Element className={classNames} data-test-id={testId} {...otherProps}>
        {children}
      </Element>
    );
  }
}

export default Subheading;
