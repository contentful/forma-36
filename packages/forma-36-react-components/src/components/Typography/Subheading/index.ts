export * from './Subheading';
export { default } from './Subheading';
