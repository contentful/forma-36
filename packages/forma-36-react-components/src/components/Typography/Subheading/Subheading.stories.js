import React from 'react';
import { storiesOf } from '@storybook/react';
import { text, selectV2 } from '@storybook/addon-knobs/react';
import { host } from 'storybook-host';
import { withInfo } from '@storybook/addon-info';

import Subheading from './Subheading';

storiesOf('Components|Typography/Subheading', module)
  .addDecorator(
    host({
      align: 'center middle',
      cropMarks: false,
    }),
  )
  .add(
    'default',
    withInfo()(() => (
      <Subheading
        extraClassNames={text('Extra Class Names', '')}
        element={selectV2(
          'Element',
          ['h1', 'h2', 'h3', 'h4', 'h5', 'h6', 'p'],
          'h1',
        )}
      >
        {text('Content', 'Subheading')}
      </Subheading>
    )),
  );
