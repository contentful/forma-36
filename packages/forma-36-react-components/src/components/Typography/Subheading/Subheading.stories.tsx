import React from 'react';
import Subheading, { SubheadingProps } from './Subheading';

export default {
  title: 'Typography/Subheading',
  component: Subheading,
  parameters: {
    propTypes: [Subheading['__docgenInfo']],
  },
  argTypes: {
    className: { control: { disable: true } },
  },
};

interface Args extends SubheadingProps {
  SubheadingText: string;
}

export const Basic = ({ SubheadingText, ...args }: Args) => (
  <Subheading {...args}>{SubheadingText}</Subheading>
);

Basic.args = {
  SubheadingText: 'Subheading',
};
