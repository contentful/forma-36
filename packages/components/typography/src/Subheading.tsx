import React, { useContext } from 'react';
import tokens from '@contentful/f36-tokens';
import { css, cx } from 'emotion';
import {
  PolymorphicComponent,
  PolymorphicComponentWithRef,
} from '@contentful/f36-core';
import type { HeadingInternalProps, HeadingProps } from './Heading';
import { Heading } from './Heading';
import { TypographyContext } from './Typography';

const DEFAULT_TAG = 'h2';

export type SubheadingProps<E extends React.ElementType> = HeadingProps<E>;

const _Subheading: PolymorphicComponentWithRef<
  HeadingInternalProps,
  typeof DEFAULT_TAG
> = ({ children, className, ...otherProps }, ref) => {
  const configuration = useContext(TypographyContext);
  return (
    <Heading
      as={DEFAULT_TAG}
      testId="cf-ui-subheading"
      marginBottom={configuration.subheading}
      className={cx(
        css({
          fontSize: tokens.fontSizeL,
          lineHeight: tokens.lineHeightL,
        }),
        className,
      )}
      {...otherProps}
      ref={ref}
    >
      {children}
    </Heading>
  );
};

export const Subheading: PolymorphicComponent<
  HeadingInternalProps,
  typeof DEFAULT_TAG
> = React.forwardRef(_Subheading);
