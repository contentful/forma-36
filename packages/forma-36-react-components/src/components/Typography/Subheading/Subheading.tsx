import React, { Component } from 'react';
import cn from 'classnames';
import styles from './Subheading.css';

import { TypographyContext } from '../Typography';

export interface SubheadingProps {
  element?: 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6' | 'p';
  className?: string;
  children?: React.ReactNode;
  testId?: string;
  style?: React.CSSProperties;
}

const defaultProps: Partial<SubheadingProps> = {
  element: 'h2',
  testId: 'cf-ui-subheading',
};

export class Subheading extends Component<SubheadingProps> {
  static defaultProps = defaultProps;

  render() {
    const { className, children, testId, element, ...otherProps } = this.props;

    const classNames = cn(styles['Subheading'], className);

    const Element = element!; // eslint-disable-line @typescript-eslint/no-non-null-assertion

    return (
      <TypographyContext.Consumer>
        {(value) => {
          return (
            <Element
              className={cn(classNames, [
                value['subheading'] &&
                  `f36-margin-bottom--${value['subheading']['spacing']}`,
              ])}
              data-test-id={testId}
              {...otherProps}
            >
              {children}
            </Element>
          );
        }}
      </TypographyContext.Consumer>
    );
  }
}

export default Subheading;
