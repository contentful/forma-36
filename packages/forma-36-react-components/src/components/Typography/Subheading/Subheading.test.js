import React from 'react';
import { shallow, mount } from 'enzyme';
import { axe } from 'jest-axe';
import Subheading from './Subheading';

it('renders the component', () => {
  const output = shallow(<Subheading>Subheading</Subheading>);

  expect(output).toMatchSnapshot();
});

it('renders the component with an additional class name', () => {
  const output = shallow(
    <Subheading extraClassNames="my-extra-class">Subheading</Subheading>,
  );

  expect(output).toMatchSnapshot();
});

it('renders the component h3', () => {
  const output = shallow(<Subheading element="h3">Subheading</Subheading>);

  expect(output).toMatchSnapshot();
});

it('has no a11y issues', async () => {
  const output = mount(<Subheading>Subheading</Subheading>).html();
  const results = await axe(output);

  expect(results).toHaveNoViolations();
});
