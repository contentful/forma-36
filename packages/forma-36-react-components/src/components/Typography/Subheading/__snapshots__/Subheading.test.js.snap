// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the component 1`] = `
<h2
  className="Subheading"
  data-test-id="cf-ui-subheading"
>
  Subheading
</h2>
`;

exports[`renders the component h3 1`] = `
<h3
  className="Subheading"
  data-test-id="cf-ui-subheading"
>
  Subheading
</h3>
`;

exports[`renders the component with an additional class name 1`] = `
<h2
  className="Subheading my-extra-class"
  data-test-id="cf-ui-subheading"
>
  Subheading
</h2>
`;
